text/0000-prefetch-hook.md

```diff
@@ -0,0 +1,82 @@
+- Start Date: 2015-10-01
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Ember's data fetching is sequential.
+Although this has advantages (one of which being that it presents a simpler programming model), the resulting latency of each request is compounded and can result in a degraded experience.
+The general solution to the "waterfall" problem of sequential requests is to execute independent requests in parallel.
+
+Automaticly inferring which requests can be made in parallel is tricky, and likely impossible to achieve while maintaining a semver compatible public API.
+Although it may not be automatic, developers have sufficient context with which they can decide what requests to make in parallel.
+
+This RFC proposes an approach that allows developers to tune the balance between sequential and parallel requests.
+
+# Motivation
+
+Currently, the `model` hook is Ember's way for routes to request data.
+However, a route's `model` hook is not entered until it's parent route has resolved.
+This is suboptimal for nested application structures.
+Child routes are forced to wait for their parents to receive data before they may make requests for their own data.
+This occurs even when the child doesn't make use of its parents' data.
+
+This RFC proposes adding a new `Route#prefetch` hook, which is used to allow routes to make requests in parallel.
+Child routes that make use of `prefetch` will resolve faster since their data can resolve at the same time as (or before) their parents'.
+The semantics and ordering of `Route`'s existing model hooks (`beforeModel`, `model`, `afterModel`) are preserved.
```

Comment:
I wonder if prefetch cause become part of the chain, essentially invoked eagerly, but its completion blocks beforeModel on that route. This may actually be least surprising. (maybe, needs thought)